icc-x-api/icc-crypto-x-api.ts

```diff
@@ -1358,39 +1347,154 @@ export class IccCryptoXApi {
+      .then(async hcp => {
+        if (autoupdateHCP) {
+          return await this.hcpartyBaseApi.modifyHealthcareParty(hcp)
+        } else {
+          return hcp
+        }
       })
   }
 
   importKeychainInBrowserFromHCP(hcpId: string): Promise<void> {
-    return this.hcpartyBaseApi.getHealthcareParty(hcpId).then(hcp => {
-      const crt = _.get(hcp.options, this.hcpPreferenceKeyEhealthCert)
-      const crtValidityDate = _.get(hcp.options, this.hcpPreferenceKeyEhealthCertDate)
-      if (crt) {
-        this.saveKeychainInBrowserLocalStorageAsBase64(hcp.id!!, crt)
+    return this.hcpartyBaseApi.getHealthcareParty(hcpId).then(async hcp => {
+      const crtCryp: ArrayBuffer = _.get(hcp.options, this.hcpPreferenceKeyEhealthCertEncrypted)
+      const crtValidityDateCryp: ArrayBuffer = _.get(
+        hcp.options,
+        this.hcpPreferenceKeyEhealthCertDateEncrypted
+      )
+
+      let crt: ArrayBuffer | null = null
+      let crtValidityDate: ArrayBuffer | null = null
+
+      let decryptionKey: CryptoKey | null = null
+      try {
+        decryptionKey = _.find(
+          await this.decryptAndImportAesHcPartyKeysForDelegators([hcp.id!], hcp.id!),
+          delegator => delegator.delegatorId === hcp.id
+        )!.key
+      } catch (e) {
+        console.error("Error while importing the AES key.")
+      }
+      if (!decryptionKey) {
+        console.error("No encryption key!")
+      }
+
+      const errors = []
+
+      if (crtCryp && decryptionKey) {
+        crt = await this.AES.decrypt(
+          decryptionKey,
+          new Uint8Array(utils.base64toArrayBuffer(btoa(utils.ua2text(new Uint8Array(crtCryp)))))
+        )
+      } else {
+        crt = _.get(hcp.options, this.hcpPreferenceKeyEhealthCert)
       }
 
-      if (crtValidityDate) {
-        this.saveKeychainValidityDateInBrowserLocalStorage(hcp.id!!, crtValidityDate)
+      if (!crt) {
+        errors.push(
+          new Error(
+            `Error while saving certificate in browser local storage! Hcp ${hcp.id} has no certificate.`
+          )
+        )
+      } else {
+        this.saveKeychainInBrowserLocalStorageAsBase64(
+          hcp.id!!,
+          utils.base64url(new Uint8Array(crt))
+        )
+      }
+
+      if (crtValidityDateCryp && decryptionKey) {
+        crtValidityDate = await this.AES.decrypt(decryptionKey, new Uint8Array(crtValidityDateCryp))
+      } else {
+        crtValidityDate = _.get(hcp.options, this.hcpPreferenceKeyEhealthCert)
       }
+
+      if (!crtValidityDate) {
+        errors.push(
+          new Error(
+            `Error while saving certificate validity date in browser local storage! Hcp ${hcp.id} has no certificate.`
+          )
+        )
+      } else {
+        this.saveKeychainValidityDateInBrowserLocalStorage(
+          hcp.id!!,
+          utils.hex2text(utils.ua2hex(crtValidityDate))
+        )
+      }
+
+      return errors.length ? Promise.reject(errors) : Promise.resolve()
```

Comment:
Why not throw an exception immediately instead of throwing an array of exceptions? We don't usually expect an array of exceptions to be thrown.

Reply:
Done.